### PR TITLE
feat: Add Streamlit Frontend

### PR DESCRIPTION
**Pull request by error9098x**
- Contains the code for frontend built with Streamlit.
- README guide includes all the information to set up the frontend as well as the optional feedback form.
- updated the .gitignore file.
- Ruff linter passed.

### CODE REVIEW

File: frontend/readme.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# FrontEnd For Streamlit
```
Comment (luarss) [inline frontend/readme.md:1]: README.md

File: frontend/readme.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+   - Export credentials as JSON and add its path in the environment variable.
+
+    ```plaintext
+    GOOGLE_CREDENTIALS_JSON=<path-to-your-google-credentials-json>
+    ```
+4. **Set the Current Version for Feedback Evaluation:**
+   - Add the current version of the feedback evaluation to the environment variables.
+    
+    ```plaintext
+    RAG_VERSION=<current-version>
```
Comment (luarss) [inline frontend/readme.md:44]: Can you automate this? This should be tied to commit number/ some sort of identifier that can be used to trace back

File: frontend/streamlit_app.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import pytz
+import os
+import ast
+from PIL import Image
+from utils.feedback import show_feedback_form
+from dotenv import load_dotenv
+
+load_dotenv()
+
+img = Image.open("assets/or_logo.png")
```
Comment (luarss) [inline frontend/streamlit_app.py:14]: `if __name__ =="__main__:`

File: .gitignore [hunk 4]
```diff
@@ -1,6 +1,8 @@
...
-data/
-backend/src/source_list.json
-.ruff_cache
-backend/src/secret.json
\ No newline at end of file
+__pycache__/
+*.pyc
+frontend/secrets.json  
+evaluation/human_evaluation/secrets.json  
+secrets.json
```
Comment (luarss) [inline .gitignore:6]: This covers all instances of `secrets.json` in the repo.

File: frontend/streamlit_app.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    })
+
+for message in st.session_state.chat_history:
+    with st.chat_message(message["role"]):
+        st.markdown(message["content"])
+
+user_input = st.chat_input("Enter your queries ...")
+
+
+def response_generator(user_input: str) -> tuple[str, str]:
```
Comment (luarss) [inline frontend/streamlit_app.py:45]: This should be above the main logic for readability

File: frontend/readme.md (hunk 2 above)
Comment (error9098x) [reply frontend/readme.md:44]: I think it would be much better if we integrate this functionality directly into our backend. This way, the RAG_VERSION can be retrieved from the backend API itself.

So, here's what I am thinking
- Use Git tags to check for any new tags when a PR is merged.
- If the tag matches a pattern like bump-version, it will automatically update the RAG_VERSION (e.g., from v0.0.1 to v0.0.2).
- This version update will happen through a commit, allowing us to track version history effectively.

If you have any suggestions do let me know.

File: .gitignore (hunk 4 above)
Comment (error9098x) [reply .gitignore:6]: Yes, it does. However, people using some other naming scheme can go under the radar as while exporting the credentials from gcp, it's usually <PROJECT_ID>.json

Comment (luarss) [reply .gitignore:6]: The practice is either you leave the last line to cover everything or just add in every instance of `secrets.json`. Not both.

Comment (luarss) [reply .gitignore:6]: If there is need for that <PROJECT_ID>.json, do `human_evaluation/*.json` or something

File: frontend/readme.md (hunk 2 above)
Comment (luarss) [reply frontend/readme.md:44]: Sounds good. If you are doing tags then also add a CHANGELOG.md file.

File: .gitignore (hunk 4 above)
Comment (error9098x) [reply .gitignore:6]: Yeah, I think this approach will be be better.

> If there is need for that <PROJECT_ID>.json, do `human_evaluation/*.json` or something

File: frontend/readme.md (hunk 2 above)
Comment (error9098x) [reply frontend/readme.md:44]: > Sounds good. If you are doing tags then also add a CHANGELOG.md file.

Yes, adding a CHANGELOG.md file will definitely make the versioning history more accessible and easier to read. I will work on implementing this.

File: frontend/readme.md (hunk 1 above)
Comment (luarss) [reply frontend/readme.md:1]: Run `git rm --cached readme.md` to remove this file.

Comment (error9098x) [reply frontend/readme.md]: Done!

File: .gitignore [hunk 6]
```diff
@@ -1,6 +1,11 @@
...
 data/
 backend/src/source_list.json
-.ruff_cache
-backend/src/secret.json
\ No newline at end of file
+backend/src/secret.json
+*.pyc
+frontend/*.json  
+evaluation/human_evaluation/*.json  
+*.json
```
Comment (luarss) [inline .gitignore:9]: Are you sure? This will lead to all jsons being ignored.

File: frontend/streamlit_app.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+                    time.sleep(0.05)
+
+                message_placeholder.markdown(response_buffer)
+            st.session_state.chat_history.append(
+                {"content": response_buffer, "role": "ai"})
+
+            if sources:
+                with st.expander("Sources:"):
+                    try:
+                        print(sources)
```
Comment (luarss) [inline frontend/streamlit_app.py:116]: Remove unnecessary `print` statements where possible.

File: frontend/streamlit_app.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            with st.chat_message("ai"):
+                message_placeholder = st.empty()
+
+                response_buffer = ""
+                for chunk in response.split(' '):
+                    response_buffer += chunk + ' '
+                    if chunk.endswith('\n'):
+                        response_buffer += ' '
+                    message_placeholder.markdown(response_buffer)
+                    time.sleep(0.05)
```
Comment (luarss) [inline frontend/streamlit_app.py:107]: Why is this needed?

Comment (luarss) [reply frontend/streamlit_app.py:107]: Are you trying to simulate streaming mode? This might help: https://github.com/langchain-ai/chat-langchain/issues/39

In fact, for long responses (100 words) you incur up to 5s delay. The right way is to enable streaming on the LLM endpoint, then process the output. What you have here is you generate the whole output, and then wait 0.05s per word. Which adds delay

File: frontend/streamlit_app.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+                            st.markdown("No valid sources found.")
+                    except (ValueError, SyntaxError) as e:
+                        st.markdown(f"Failed to parse sources: {e}")
+
+    question_dict = {
+        interaction["content"]: i
+        for i, interaction in enumerate(st.session_state.chat_history)
+        if interaction["role"] == "user"
+    }
+    if question_dict and os.getenv("FEEDBACK_SHEET_ID"):
```
Comment (luarss) [inline frontend/streamlit_app.py:140]: if `FEEDBACK_SHEET_ID` is necessary, make it a if condition within the `if question_dict` logic. Otherwise you will just not display anything if its not provided

File: frontend/utils/feedback.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    Returns:
+    - str: The title of the sheet with the given GID.
+    """
+    sheets = spreadsheet_metadata["sheets"]
+    for sheet in sheets:
+        if sheet["properties"]["sheetId"] == int(gid):
+            return sheet["properties"]["title"]
+    return None
+
+def format_sources(sources) -> str:
```
Comment (luarss) [inline frontend/utils/feedback.py:27]: types?

File: frontend/utils/feedback.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    - sources (list): List of source URLs.
+
+    Returns:
+    - str: Formatted sources string.
+    """
+    if isinstance(sources, list):
+        return "\n".join(sources)
+    return str(sources)
+
+def format_context(context) -> str:
```
Comment (luarss) [inline frontend/utils/feedback.py:41]: types?

File: frontend/streamlit_app.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+    user_input = st.chat_input("Enter your queries ...")
+
+    if user_input:
+        st.session_state.chat_history.append({"content": user_input, "role": "user"})
+
+        with st.chat_message("user"):
+            st.markdown(user_input)
+
+        response, sources = response_generator(user_input)
```
Comment (luarss) [inline frontend/streamlit_app.py:94]: Have we implemented conversational memory? i.e. is the assistant able to recall past questions in the conversation history?

File: frontend/streamlit_app.py (hunk 8 above)
Comment (error9098x) [reply frontend/streamlit_app.py:107]: This isn't needed; it's just a visual thing. It's inefficient, as you said. It was initially made with @palaniappan-r 's endpoint structure in mind. Now, all the features such as human evaluation, auto evaluation, feedback, and frontend are dependent on the current API structure. For streaming to happen directly from LangChain, I will need to coordinate with Pal and create a separate /stream endpoint and test the possibility of it.

File: frontend/streamlit_app.py (hunk 12 above)
Comment (error9098x) [reply frontend/streamlit_app.py:94]: Not yet, @palaniappan-r  wanted to do that in the backend.

File: frontend/streamlit_app.py (hunk 9 above)
Comment (error9098x) [reply frontend/streamlit_app.py:140]: That's what I intended. If the user hasn't set up the feedback sheet ID, there's no need to display a feedback button and the feedback side page.

File: .gitignore (hunk 6 above)
Comment (luarss) [reply .gitignore:9]: Yes `/*.json` is the way to ignore all root dir json files. The way `*.json` works is by globbing, so it matches everything that ends with `.json`. 

https://github.com/kenmueller/gitignore

File: frontend/readme.md (hunk 2 above)
Comment (luarss) [reply frontend/readme.md]: Ok, let's target this in future PRs.

**Comment by luarss** [comment]
@error9098x  Please fix the merge conflicts

**Review by luarss — APPROVED** [review]

**Review by luarss — APPROVED** [review]

**Review by luarss — APPROVED** [review]
